pkg/kubectl/cmd/edit.go

```diff
@@ -396,56 +395,50 @@ func visitToPatch(originalObj runtime.Object, updates *resource.Info, mapper met
 
 		}
 
-		originalSerialization, err := runtime.Encode(encoder, currOriginalObj)
+		ok, err := checkAPiVersionKindName(currOriginalObj, info.Object)
 		if err != nil {
 			return err
 		}
-		editedSerialization, err := runtime.Encode(encoder, info.Object)
-		if err != nil {
-			return err
-		}
-
-		// compute the patch on a per-item basis
-		// use strategic merge to create a patch
-		originalJS, err := yaml.ToJSON(originalSerialization)
-		if err != nil {
-			return err
-		}
-		editedJS, err := yaml.ToJSON(editedSerialization)
-		if err != nil {
-			return err
-		}
-
-		if reflect.DeepEqual(originalJS, editedJS) {
-			// no edit, so just skip it.
-			cmdutil.PrintSuccess(mapper, false, out, info.Mapping.Resource, info.Name, false, "skipped")
-			return nil
-		}
-
-		preconditions := []strategicpatch.PreconditionFunc{strategicpatch.RequireKeyUnchanged("apiVersion"),
-			strategicpatch.RequireKeyUnchanged("kind"), strategicpatch.RequireMetadataKeyUnchanged("name")}
-		patch, err := strategicpatch.CreateTwoWayMergePatch(originalJS, editedJS, currOriginalObj, preconditions...)
-		if err != nil {
-			glog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
-			if strategicpatch.IsPreconditionFailed(err) {
-				return preservedFile(nil, file, errOut)
-			}
-			return err
+		if !ok {
+			return goerrors.New("At least one of apiVersion, kind and name was changed")
```

Comment:
Not very clear, try something like "The following fields must not be changed: apiVersion, kind, and metadata.name".